### PR TITLE
Podcasts 

### LINKED ISSUE #90: Can't play Podcasts
<a href="https://github.com/michaelherger"></img></a> **Issue by [michaelherger](https://github.com/michaelherger)**
_Friday Sep 22, 2017 at 08:21 GMT_
_Originally opened as https://github.com/plietar/librespot/issues/250_

----

I can't play Spotify's Podcasts on librespot. When playing in the desktop app, librespot based clients would be visible in the "Devices Available" list, but greyed out. And when a librespot based client is active, I can't start playback of podcasts. 

Trying to play a podcast to librespot wouldn't log anything in librespot, not even in verbose mode. I must assume that Spotify wouldn't send any data for whatever reason.

### PR DESCRIPTION
**Pull request by ashthespy**
Fixes #90
Ported an old attempt for podcast support, and added suggestions from #339 and #377 

There are still a few protobuf fields of the endpoint that I haven't been able to reverse fully, but these are not important for playback.  

I am unsure how to proceed with respect to the `SpotifyID` . 
1. Here I just add a new `SpotifyAudioType` field to the structure.  An other option is to convert to it an enum -  (https://github.com/librespot-org/librespot/pull/339#issuecomment-506489625). But both these options don't solve the base issue of 2.
2. Should we move towards storing more metadata (Track/Album/Show/Episode etc) in `SpotifyID`? 

EDIT: ~Ah, I should probably rename some proto defs to match https://github.com/librespot-org/librespot/issues/308~

I should mention that this isn't a comprehensive implementation, stuff missing includes 
- ~Track episode played state~
  - ~Hopefully~ this is done server side, ~needs some more testing~  
- Consumption order 
- Playback speed

### CODE REVIEW

File: protocol/files.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// Autogenerated by build.rs
+pub const FILES: &'static [(&'static str, u32)] = &[
+    ("proto/authentication.proto", 2098196376),
+    ("proto/keyexchange.proto", 451735664),
+    ("proto/mercury.proto", 709993906),
+    ("proto/metadata.proto", 334477813),
+    ("proto/pubsub.proto", 2686584829),
+    ("proto/spirc.proto", 1587493382),
+];
```
Comment (willstott101) [inline protocol/files.rs]: This should be gone in current master?

Comment (ashthespy) [reply protocol/files.rs]: True, old habits!

**Comment by willstott101** [comment]
This way looks fine to me, but if we have much more information it should probably be a `SpotifyObject` or smth which contains a `SpotifyID`? Rather than setting up SpotifyID to grow.

I was just exploring the closest pattern I could think of in Rust to match the java implementation.

**Comment by ashthespy** [comment]
> 
> 
> This way looks fine to me, but if we have much more information it should probably be a `SpotifyObject` or smth which contains a `SpotifyID`? Rather than setting up SpotifyID to grow.
> 
> I was just exploring the closest pattern I could think of in Rust to match the java implementation.

That might be a better option indeed, cause other than for metadata (Track/Album/Artist/Show/Episode), the `player` needs only the `gid` to fetch files. Only `spirc` needs to be context/URI aware.

File: core/src/spotify_id.rs [hunk 2]
```diff
@@ -53,11 +68,22 @@ impl SpotifyId {
         let high = BigEndian::read_u64(&data[0..8]);
         let low = BigEndian::read_u64(&data[8..16]);
 
-        Ok(SpotifyId(u128::from_parts(high, low)))
+        Ok(SpotifyId::as_track(u128::from_parts(high, low)))
+    }
+
+    pub fn from_uri(uri: &str) -> Result<SpotifyId, SpotifyIdError> {
+        let parts = uri.split(":").collect::<Vec<&str>>();
+        if uri.contains(":show:") || uri.contains(":episode:") {
```
Comment (willstott101) [inline core/src/spotify_id.rs]: Looks like `spotify:collection:podcasts:episodes` is a valid form too :)

https://github.com/librespot-org/librespot/pull/339#issuecomment-505139197

https://github.com/librespot-org/librespot/pull/339/commits/25bfc0582da61e29675474b7c516a1c6de45b289

Comment (willstott101) [reply core/src/spotify_id.rs]: Can we also only have one implementation of this check? Seems to be in two places.

**Comment by devgianlu** [comment]
> That might be a better option indeed, cause other than for metadata (Track/Album/Artist/Show/Episode), the player needs only the gid to fetch files. Only spirc needs to be context/URI aware.

I think that's how it should be. In librespot-java the context URI is not important at all because the only important thing is the URI of the track/episode which is pretty standard.

File: core/src/spotify_id.rs (hunk 2 above)
Comment (ashthespy) [reply core/src/spotify_id.rs]: Pushed some refactoring that should explain why I wasn't seeing the issue with `collection:podcasts`

**Comment by ashthespy** [comment]
Rebased this on dev after the merged changes in `spotify_id` - give it a whirl! :-)

**Comment by ashthespy** [comment]
If there are no issues from anyone, will go ahead and merge this in tonight..

**Pull request merged**